lib/nerves/artifact/resolvers/github_api.ex

```diff
@@ -2,143 +2,132 @@ defmodule Nerves.Artifact.Resolvers.GithubAPI do
   @moduledoc false
   @behaviour Nerves.Artifact.Resolver
 
-  alias Nerves.Utils
-  alias Nerves.Utils.HTTPClient
+  alias Nerves.{Utils, Utils.HTTPClient}
 
   @base_url "https://api.github.com/"
-  @required_opts [:username, :token, :tag]
+
+  defstruct artifact_name: nil,
+            base_url: @base_url,
+            headers: [],
+            http_client: HTTPClient,
+            http_pid: nil,
+            public?: false,
+            opts: [],
+            repo: nil,
+            tag: "",
+            token: "",
+            url: nil,
+            username: ""
 
   @impl Nerves.Artifact.Resolver
   def get({org_proj, opts}) do
-    case validate_opts(opts) do
-      {:ok, opts} ->
-        fetch_artifact(org_proj, opts)
+    opts =
+      %{struct(__MODULE__, opts) | opts: opts, repo: org_proj}
+      |> maybe_adjust_token()
+      |> add_http_opts()
+      |> maybe_start_http()
 
-      {:error, missing} ->
-        missing = Enum.join(missing, ", ")
-        artifact_name = opts[:artifact_name]
+    result = fetch_artifact(opts)
 
-        error = """
-        Skipping: #{artifact_name}
-        Required options missing: [#{missing}]
-        """
+    opts.http_client.stop(opts.http_pid)
 
-        {:error, error}
-    end
+    result
   end
 
-  defp fetch_artifact(org_proj, opts) do
-    artifact_name = opts[:artifact_name]
-    {:ok, http_pid} = HTTPClient.start_link()
-    token = get_token(opts, artifact_name)
-    username = get_username(opts, artifact_name)
-    tag = get_tag(opts, artifact_name)
-
-    credentials = Base.encode64(username <> ":" <> token)
-    auth_header = {"Authorization", "Basic " <> credentials}
-    accept_header = {"Accept", "application/octet-stream"}
-    base_url = opts[:base_url] || @base_url
-
-    url = Path.join([base_url, "repos", org_proj, "releases", "tags", tag])
-
-    Nerves.Utils.Shell.info("  Downloading artifacts from #{url}")
-
-    url = URI.encode(url)
-
-    result =
-      with {:ok, data} <- HTTPClient.get(http_pid, url, headers: [auth_header], progress?: false),
-           %{"assets" => assets} <- Utils.json_decode(data),
-           {:ok, url} <- get_asset_url(assets, artifact_name) do
-        HTTPClient.get(http_pid, url, headers: [auth_header, accept_header])
+  defp add_http_opts(opts) do
+    headers =
+      if opts.public? do
+        []
+      else
+        credentials = Base.encode64(opts.username <> ":" <> opts.token)
+        [{"Authorization", "Basic " <> credentials}]
       end
 
-    Nerves.Utils.HTTPClient.stop(http_pid)
-    result
+    %{
+      opts
+      | headers: headers,
+        url: Path.join([opts.base_url, "repos", opts.repo, "releases", "tags", opts.tag])
+    }
   end
 
-  @spec validate_opts(Keyword.t()) :: {:ok | :error, Keyword.t()}
-  def validate_opts(opts) do
-    keys = Keyword.keys(opts)
+  defp maybe_adjust_token(opts) do
+    token = System.get_env("GITHUB_TOKEN") || System.get_env("GH_TOKEN")
```

Comment:
> Supports `GITHUB_TOKEN`

FYI as I've set up Nerves projects with GitHub Actions instead of CircleCI, it's prevented storing secrets that start with `GITHUB` because it's a reserved word. I got around this by changing the secret to something like `ARTIFACT_TOKEN` and updating the CI config. Just thought I'd mention it while I see this code changing. Also maybe that's what `GH_TOKEN` is for and I just didn't see it earlier...

Reply:
Interesting. `GH_TOKEN` was to match the [GitHub CLI supported environment variables](https://cli.github.com/manual/gh_help_environment). I thought `GITHUB_*` would take precedence, but looks like it is the other way around. We could switch that.

I would say that we leave it as is rather than try to make up new environment variables. This is more of a convenience and to potentially provide more helpful errors. If some special env needs to be used, I would say that should go in your Nerves systems `nerve_package()` function when defining the artifact:

```elixir
def nerves_package() do
  [
    artifact_sites: [
      {:github_api, "my_org/my_system/, token: System.get_env("MY_ARTIFACT_TOKEN")}
    ]
  ]
end
```

Reply:
>I would say that we leave it as is

👍 Sure, no problem. I've been using a workaround similar to what you posted and it was also easy enough to update the CI config for it.